source/features/select-notifications.tsx

```diff
@@ -0,0 +1,176 @@
+import React from 'dom-chef';
+import select from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+import {
+	CheckCircleIcon,
+	CheckIcon,
+	DotFillIcon,
+	DotIcon,
+	GitMergeIcon,
+	GitPullRequestDraftIcon,
+	GitPullRequestIcon,
+	IssueOpenedIcon,
+	XCircleIcon
+} from '@primer/octicons-react';
+
+import features from '.';
+
+const filters = {
+	'Pull requests': ':is(.octicon-git-pull-request, .octicon-git-pull-request-closed, .octicon-git-pull-request-draft, .octicon-git-merge)',
+	Issues: ':is(.octicon-issue-opened, .octicon-issue-closed)',
+	Open: ':is(.octicon-issue-opened, .octicon-git-pull-request)',
+	Closed: ':is(.octicon-issue-closed, .octicon-git-pull-request-closed)',
+	Draft: '.octicon-git-pull-request-draft',
+	Merged: '.octicon-git-merge',
+	Read: '.notification-read',
+	Unread: '.notification-unread'
+};
+
+type Filter = keyof typeof filters;
+type Category = 'Type' | 'Status' | 'Read';
+
+function resetFilters({target}: Event): void {
+	for (const checkbox of select.all('input[type="checkbox"]:checked', target as Element)) {
+		checkbox.checked = false;
+		checkbox.parentElement!.parentElement!.setAttribute('aria-checked', 'false');
+	}
+}
+
+function handleSelection({target}: Event): void {
+	const selectedFilters = select.all('[aria-checked="true"]', target as Element);
+	const selectAllCheckbox = select('input[type="checkbox"].js-notifications-mark-all-prompt')!;
+	// Reset the "Select all" checkbox
+	if (selectAllCheckbox.checked) {
+		selectAllCheckbox.dispatchEvent(new MouseEvent('click'));
+	}
+
+	if (selectedFilters.length > 0) {
+		const activeFilters: Record<Category, string[]> = {
+			Type: [],
+			Status: [],
+			Read: []
+		};
+		for (const selectedFilter of selectedFilters) {
+			activeFilters[selectedFilter.dataset.category as Category].push(filters[selectedFilter.dataset.filter as Filter]);
+		}
+
+		for (const notification of select.all('.notifications-list-item')) {
+			let isNotificationSelected = true;
+			for (const [category, categoryFilters] of Object.entries(activeFilters)) {
+				if (category === 'Read') {
+					if (categoryFilters.length === 1 && !notification.matches(categoryFilters[0])) {
+						isNotificationSelected = false;
+						break;
+					}
+
+					continue;
+				}
+
+				if (categoryFilters.length > 0 && !select.exists(categoryFilters, notification)) {
+					isNotificationSelected = false;
+					break;
+				}
+			}
+
+			// Make excluded notifications unselectable
+			if (!isNotificationSelected) {
+				select('.js-notification-bulk-action-check-item', notification)!.removeAttribute('data-check-all-item');
+			}
+		}
```

Comment:
I still think this loop can/should be simplified.

Since this is ignoring unselected items, just query the items you're looking for and loop them specifically. I think that even if they're 3 separate loops they're more readable and less error-prone than several nested loops and conditions.

Improvable pseudo-code:

```js
const types = new FormData(formElement).getAll('Type').map(type => filters[type]).join('') // '' means "AND"
for (const type of types) {
	// select it
}

const status = new FormData(formElement).getAll('Status').map(type => filters[type]).join(',') // ',' means "OR"
for (const status of status) {
	// select it
}

// etc
```

Reply:
> there's no way to express that in a selector.

Why not? Isn't technically each filter applied on top of the previous selection?

1. Select PRs
2. Out of the selected PRs, select closed or merged
3. Out of the selected closed or merged PRs, select unread

Reply:
The current code looks good though@